File: kernel/src/diagnostics.ts
import { evaluateRelation, type GeometricRelation } from "./relations.js";
import type { Constraint, DesignProgramSnapshot } from "./core.js";

export interface DiagnosticClassification {
  readonly redundantConstraintIds:readonly string[];
  readonly contradictoryConstraintIds:readonly string[];
  readonly contradictoryRelationIndexes:readonly number[];
  readonly warnings:readonly string[];
}

function canonical(value:unknown):string{if(value===null||typeof value!=="object")return JSON.stringify(value);if(Array.isArray(value))return`[${value.map(canonical).join(",")}]`;const record=value as Record<string,unknown>;return`{${Object.keys(record).sort().map(key=>`${JSON.stringify(key)}:${canonical(record[key])}`).join(",")}}`;}
function isConstraint(value:unknown):value is Constraint{return!!value&&typeof value==="object"&&typeof (value as {readonly kind?:unknown}).kind==="string";}

export function classifyConstraintSystem(snapshot:DesignProgramSnapshot,relations:readonly GeometricRelation[]=[]):DiagnosticClassification{
  const seen=new Map<string,string>(),redundant:string[]=[],contradictory:string[]=[],warnings:string[]=[];
  for(const item of snapshot.constraints){const key=canonical(item.constraint);const previous=seen.get(key);if(previous){redundant.push(item.id);warnings.push(`Constraint ${item.id} duplicates ${previous}`);}else seen.set(key,item.id);}
  const byEntity=new Map<string,Constraint[]>();
  for(const item of snapshot.constraints){if(!isConstraint(item.constraint))continue;const c=item.constraint;const id=c.kind==="horizontal"||c.kind==="vertical"||c.kind==="fixed"?c.entityId:c.firstGeometryId;const list=byEntity.get(id)??[];list.push(c);byEntity.set(id,list);}
  for(const [id,list] of byEntity){const hasH=list.some(c=>c.kind==="horizontal"),hasV=list.some(c=>c.kind==="vertical");if(hasH&&hasV){for(const item of snapshot.constraints)if((item.constraint.kind==="horizontal"||item.constraint.kind==="vertical")&&item.constraint.entityId===id)contradictory.push(item.id);warnings.push(`Geometry ${id} cannot satisfy both horizontal and vertical constraints`);}
    const distances=list.filter((c):c is Extract<Constraint,{kind:"distance"}>=>c.kind==="distance"&&!c.secondGeometryId&&c.firstGeometryId===id);const values=new Set(distances.map(c=>c.value));if(values.size>1){for(const item of snapshot.constraints)if(item.constraint.kind==="distance"&&item.constraint.firstGeometryId===id&&!item.constraint.secondGeometryId)contradictory.push(item.id);warnings.push(`Geometry ${id} has incompatible scalar distance constraints`);}}
  const contradictoryRelationIndexes:number[]=[];
  for(let i=0;i<relations.length;i+=1){for(let j=i+1;j<relations.length;j+=1){const a=relations[i]!,b=relations[j]!;if(a.kind==="parallel"&&b.kind==="perpendicular"&&sameUnorderedPair(a.firstGeometryId,a.secondGeometryId,b.firstGeometryId,b.secondGeometryId))contradictoryRelationIndexes.push(i,j);if(a.kind==="perpendicular"&&b.kind==="parallel"&&sameUnorderedPair(a.firstGeometryId,a.secondGeometryId,b.firstGeometryId,b.secondGeometryId))contradictoryRelationIndexes.push(i,j);if(a.kind==="radius"&&b.kind==="radius"&&a.geometryId===b.geometryId&&a.value!==b.value)contradictoryRelationIndexes.push(i,j);if(a.kind==="diameter"&&b.kind==="diameter"&&a.geometryId===b.geometryId&&a.value!==b.value)contradictoryRelationIndexes.push(i,j);}}
  return Object.freeze({redundantConstraintIds:Object.freeze([...new Set(redundant)]),contradictoryConstraintIds:Object.freeze([...new Set(contradictory)]),contradictoryRelationIndexes:Object.freeze([...new Set(contradictoryRelationIndexes)].sort((a,b)=>a-b)),warnings:Object.freeze(warnings)});
}

function sameUnorderedPair(a:string,b:string,c:string,d:string):boolean{return(a===c&&b===d)||(a===d&&b===c);}

export function relationConsistency(snapshot:DesignProgramSnapshot,relations:readonly GeometricRelation[]):readonly string[]{const out:string[]=[];for(let i=0;i<relations.length;i+=1){try{const r=evaluateRelation(snapshot,relations[i]!);if(!Number.isFinite(r.scaledNorm))out.push(`Relation ${i+1} produced a non-finite residual`);}catch(error){out.push(`Relation ${i+1} invalid: ${error instanceof Error?error.message:String(error)}`);}}return Object.freeze(out);}

File: kernel/math/diagnostics.rs
use super::snapshot::{Constraint, Relation, SemanticSnapshot};

#[derive(Clone, Debug, PartialEq)]
pub struct DiagnosticClassification {
    pub redundant_constraint_ids: Vec<String>,
    pub contradictory_constraint_ids: Vec<String>,
    pub contradictory_relation_indexes: Vec<usize>,
    pub warnings: Vec<String>,
}

pub fn classify_constraints(snapshot: &SemanticSnapshot) -> DiagnosticClassification {
    let mut result = DiagnosticClassification {
        redundant_constraint_ids: Vec::new(),
        contradictory_constraint_ids: Vec::new(),
        contradictory_relation_indexes: Vec::new(),
        warnings: Vec::new(),
    };
    for i in 0..snapshot.constraints.len() {
        for j in 0..i {
            if snapshot.constraints[i].1 == snapshot.constraints[j].1 {
                result
                    .redundant_constraint_ids
                    .push(snapshot.constraints[i].0.clone());
                result.warnings.push(format!(
                    "Constraint {} duplicates {}",
                    snapshot.constraints[i].0, snapshot.constraints[j].0
                ));
            }
        }
    }
    for (id, c) in &snapshot.constraints {
        if let Constraint::Horizontal { entity_id } = c {
            if snapshot
                .constraints
                .iter()
                .any(|(_, x)| matches!(x,Constraint::Vertical{entity_id:other} if other==entity_id))
            {
                result.contradictory_constraint_ids.push(id.clone());
            }
        }
        if let Constraint::Vertical { entity_id } = c {
            if snapshot.constraints.iter().any(
                |(_, x)| matches!(x,Constraint::Horizontal{entity_id:other} if other==entity_id),
            ) {
                result.contradictory_constraint_ids.push(id.clone());
            }
        }
    }
    for i in 0..snapshot.relations.len() {
        for j in 0..i {
            let a = &snapshot.relations[i].1;
            let b = &snapshot.relations[j].1;
            if conflicting_relations(a, b) {
                result.contradictory_relation_indexes.extend([j, i]);
            }
        }
    }
    result.contradictory_constraint_ids.sort();
    result.contradictory_constraint_ids.dedup();
    result.contradictory_relation_indexes.sort();
    result.contradictory_relation_indexes.dedup();
    result
}

fn pair(a: &str, b: &str, c: &str, d: &str) -> bool {
    (a == c && b == d) || (a == d && b == c)
}
fn conflicting_relations(a: &Relation, b: &Relation) -> bool {
    match (a, b) {
        (
            Relation::Parallel {
                first_geometry_id: a1,
                second_geometry_id: a2,
            },
            Relation::Perpendicular {
                first_geometry_id: b1,
                second_geometry_id: b2,
            },
        )
        | (
            Relation::Perpendicular {
                first_geometry_id: a1,
                second_geometry_id: a2,
            },
            Relation::Parallel {
                first_geometry_id: b1,
                second_geometry_id: b2,
            },
        ) => pair(a1, a2, b1, b2),
        (
            Relation::Radius {
                geometry_id: a,
                value: x,
            },
            Relation::Radius {
                geometry_id: b,
                value: y,
            },
        ) if a == b => x != y,
        (
            Relation::Diameter {
                geometry_id: a,
                value: x,
            },
            Relation::Diameter {
                geometry_id: b,
                value: y,
            },
        ) if a == b => x != y,
        _ => false,
    }
}
